java/memory/src/main/java/org/apache/arrow/memory/BufferLedger.java

```diff
@@ -0,0 +1,515 @@
+   */
+  @Override
+  public boolean release() {
+    return release(1);
+  }
+
+  /**
+   * Decrement the ledger's reference count for the associated underlying
+   * memory chunk. If the reference count drops to 0, it implies that
+   * no ArrowBufs managed by this reference manager need access to the memory
+   * chunk. In that case, the ledger should inform the allocation manager
+   * about releasing its ownership for the chunk. Whether or not the memory
+   * chunk will be released is something that {@link AllocationManager} will
+   * decide since tracks the usage of memory chunk across multiple reference
+   * managers and allocators.
+   * @param decrement amount to decrease the reference count by
+   * @return true if the new ref count has dropped to 0, false otherwise
+   */
+  @Override
+  public boolean release(int decrement) {
+    Preconditions.checkState(decrement >= 1, "ref count decrement should be greater than or equal to 1");
+    // decrement the ref count
+    final int refCnt = decrement(decrement);
+    if (BaseAllocator.DEBUG) {
+      historicalLog.recordEvent("release(%d). original value: %d", decrement, refCnt + decrement);
+    }
+    // the new ref count should be >= 0
+    Preconditions.checkState(refCnt >= 0, "RefCnt has gone negative");
+    return refCnt == 0;
+  }
+
+  /**
+   * Decrement the ledger's reference count for the associated underlying
+   * memory chunk. If the reference count drops to 0, it implies that
+   * no ArrowBufs managed by this reference manager need access to the memory
+   * chunk. In that case, the ledger should inform the allocation manager
+   * about releasing its ownership for the chunk. Whether or not the memory
+   * chunk will be released is something that {@link AllocationManager} will
+   * decide since tracks the usage of memory chunk across multiple reference
+   * managers and allocators.
+   *
+   * @param decrement amount to decrease the reference count by
+   * @return the new reference count
+   */
+  private int decrement(int decrement) {
+    allocator.assertOpen();
+    final int outcome;
+    synchronized (allocationManager) {
+      outcome = bufRefCnt.addAndGet(-decrement);
+      if (outcome == 0) {
+        lDestructionTime = System.nanoTime();
+        // refcount of this reference manager has dropped to 0
+        // inform the allocation manager that this reference manager
+        // no longer holds references to underlying memory
+        allocationManager.release(this);
+      }
+    }
+    return outcome;
+  }
+
+  /**
+   * Increment the ledger's reference count for associated
+   * underlying memory chunk by 1
+   */
+  @Override
+  public void retain() {
+    retain(1);
+  }
+
+  /**
+   * Increment the ledger's reference count for associated
+   * underlying memory chunk by the given amount
+   * @param increment amount to increase the reference count by
+   */
+  @Override
+  public void retain(int increment) {
```

Comment:
Again here, could it be useful to return the new reference count?

Reply:
May be. I just kept the way it was in previous implementation of ArrowBuf.